src/internal/lokiutil/repair.go

```diff
@@ -0,0 +1,102 @@
+package lokiutil
+
+// This file repairs log lines from broken Loki configurations.  Learn about the log formats here:
+// https://github.com/kubernetes/design-proposals-archive/blob/main/node/kubelet-cri-logging.md
+
+import (
+	"encoding/json"
+	"strings"
+)
+
+func tryDocker(input map[string]any) (result string) {
+	for k, v := range input {
+		switch k {
+		case "log":
+			if s, ok := v.(string); ok {
+				result = s
+				continue
+			}
+			// Malformed.  "log" must be a string.
+			return ""
+		case "stream", "time":
+			continue
+		default:
+			// Docker logs contain only "log", "stream", and "time" keys.  time
+			// and stream are treated as optional.  If we see a different key,
+			// then this isn't Docker log JSON.
+			return ""
+		}
+	}
+	return
+}
+
+func tryCRI(input string) (result string) {
+	parts := strings.SplitN(input, " ", 4)
+	if len(parts) != 4 {
+		return input
+	}
+	time, stream, flag, msg := parts[0], parts[1], parts[2], parts[3]
+	switch flag {
+	case "P", "F":
+	default:
+		return input
+	}
+	switch stream {
+	case "stdout", "stderr":
+	default:
+		return input
+	}
+	for _, r := range time {
+		switch r {
+		case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'T', 'Z', ':', '.', '+', '-':
+			// We don't do a full parse of the time; if it contains only valid
+			// RFC3339Nano bytes, then we assume it's valid enough to continue testing
```

Comment:
Why not parse it as `RFC3339Nano`? Performance?

Reply:
We would just throw the result away, so I don't think it's worth it.

Reply:
But then you would know if it really is a time or not — if it’s an error, then you know that it’s not a valid CRI log.

Reply:
It's 10% slower, but I made this change.